util/autogen_md.py

```diff
@@ -0,0 +1,86 @@
+#!/usr/bin/env python3
+# Copyright lowRISC contributors.
+# Licensed under the Apache License, Version 2.0, see LICENSE for details.
+# SPDX-License-Identifier: Apache-2.0
+"""Go through all markdown files and autogenerate content based on commands
+recorded in the file.
+"""
+
+import sys
+from pathlib import Path
+import re
+import subprocess
+import argparse
+import os
+from typing import Callable
+
+START_MARKER_PATTERN = re.compile(r"\n<!--\s*BEGIN\s*AUTOGEN\s*(?P<cmd>.+)\s*-->\n")
+END_MARKER_PATTERN = re.compile(r"\n<!--\s*END\s*AUTOGEN\s*-->\n")
+
+
+def apply_to_all_files(path: Path, fn: Callable[[Path], None]):
+    for child in path.iterdir():
+        if child.is_file():
+            fn(child)
+        if child.is_dir():
+            apply_to_all_files(child, fn)
+
+
+def autogen_rewrite_md(filepath: Path, dry_run: bool):
+    # only consider .md files
+    if not filepath.name.endswith(".md"):
+        return
+    content = filepath.read_text()
+    modified = False
+    pos = 0
+    while True:
+        # search next start marker
+        match_start = START_MARKER_PATTERN.search(content, pos)
+        if match_start is None:
+            # no more replacements to do
+            break
+        cmd = match_start.group('cmd').strip()
+        # search end marker after the start marker
+        match_end = END_MARKER_PATTERN.search(content, pos)
+        if match_end is None:
+            sys.exit(f"Error in {filepath}: start marker with command '{cmd}' without end marker")
+
+        print(f"In {filepath}: running '{cmd}'")
+        if dry_run:
+            # don't run
+            pos = match_end.end(0)
+            continue
+        res = subprocess.run(cmd, shell=True, capture_output=True)
```

Comment:
We will want to be careful where/how we execute this script, since we are potentially passing unsanitized inputs to the shell. Specifically, if someone submits a PR with an AUTOGEN block in it, any command could get executed (i.e., shell injection). If this only runs post-merge, then we could do some sanitization steps pre-merge or have some CI checks/bots that flag these for specific reviewer approval. Just wanted to point this out.

Reply:
Yes, good to point out. It would be good to make it clear to reviewers that these commands are executed by linters